File: sdks/dotnet/src/Mobiscroll.Connect/Internal/QueryStringBuilder.cs
using System.Collections.Generic;
using System.Net;
using System.Text;

namespace Mobiscroll.Connect.Internal;

/// <summary>
/// URL-encoded query string builder that matches the wire format of the Node and PHP SDKs:
/// booleans serialize as the literal strings "true"/"false" (not 1/0), and complex values
/// are expected to be pre-serialized (e.g. JSON for calendarIds) by the caller.
/// </summary>
internal sealed class QueryStringBuilder
{
    private readonly List<KeyValuePair<string, string>> _items = new();

    public QueryStringBuilder Add(string key, string? value)
    {
        if (value is not null)
        {
            _items.Add(new KeyValuePair<string, string>(key, value));
        }
        return this;
    }

    public QueryStringBuilder Add(string key, int? value)
    {
        if (value.HasValue)
        {
            _items.Add(new KeyValuePair<string, string>(key, value.Value.ToString(System.Globalization.CultureInfo.InvariantCulture)));
        }
        return this;
    }

    public QueryStringBuilder Add(string key, bool? value)
    {
        if (value.HasValue)
        {
            _items.Add(new KeyValuePair<string, string>(key, value.Value ? "true" : "false"));
        }
        return this;
    }

    public bool IsEmpty => _items.Count == 0;

    public override string ToString()
    {
        if (_items.Count == 0)
        {
            return string.Empty;
        }

        var sb = new StringBuilder();
        for (var i = 0; i < _items.Count; i++)
        {
            if (i > 0)
            {
                sb.Append('&');
            }
            sb.Append(WebUtility.UrlEncode(_items[i].Key));
            sb.Append('=');
            sb.Append(WebUtility.UrlEncode(_items[i].Value));
        }
        return sb.ToString();
    }
}

File: sdks/java/connect-sdk/src/main/java/com/mobiscroll/connect/internal/QueryStringBuilder.java
package com.mobiscroll.connect.internal;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Builds URL query strings. Booleans are encoded as the strings {@code "true"} / {@code "false"}
 * to match the wire format used by the Node, PHP, and .NET SDKs.
 */
public final class QueryStringBuilder {

    private final Map<String, List<String>> params = new LinkedHashMap<>();

    public QueryStringBuilder add(String key, Object value) {
        if (value == null) {
            return this;
        }
        if (value instanceof Boolean) {
            addOne(key, ((Boolean) value) ? "true" : "false");
        } else if (value instanceof OffsetDateTime) {
            addOne(key, ((OffsetDateTime) value).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
        } else if (value instanceof Collection) {
            for (Object v : (Collection<?>) value) {
                if (v != null) {
                    addOne(key, String.valueOf(v));
                }
            }
        } else {
            addOne(key, String.valueOf(value));
        }
        return this;
    }

    private void addOne(String key, String value) {
        params.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
    }

    public boolean isEmpty() {
        return params.isEmpty();
    }

    /** @return the encoded query string without a leading {@code ?}. */
    public String build() {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, List<String>> e : params.entrySet()) {
            for (String v : e.getValue()) {
                if (sb.length() > 0) {
                    sb.append('&');
                }
                sb.append(encode(e.getKey())).append('=').append(encode(v));
            }
        }
        return sb.toString();
    }

    private static String encode(String s) {
        try {
            return URLEncoder.encode(s, StandardCharsets.UTF_8.name());
        } catch (UnsupportedEncodingException e) {
            throw new IllegalStateException(e);
        }
    }
}
